### PR TITLE
perf(disagg): pipeline mm_inputs processing with GPU forward in prefill

### PR DESCRIPTION
**Pull request by Jumiar**
## Motivation
Overlap heavy mm_inputs processing (from_processor_output + SHM tensor
  clone) with GPU forward execution by running it in a background
  ThreadPoolExecutor.

  Pipeline structure (event_loop_overlap_disagg_prefill):
    1. Collect bg mm results from previous iteration, dispatch requests
    2. recv_requests (Gloo broadcast syncs all ranks)
    3. Submit current mm_inputs to bg thread
    4. Schedule next batch
    5. GPU forward (bg thread runs concurrently)
    6. Process last batch result

  - scheduler.py: add SHM lifecycle state and pipelined mm processing
    helpers (_submit_mm_processing, _collect_mm_results_and_dispatch,
    generation-based deferred SHM unlink); handle_generate_request
    checks _bg_mm_result before falling back to synchronous
    _get_multimodal_inputs.
  - mm_utils.py: add ShmPointerMMData.read_tensor()/unlink_shm() and
    materialize_shm_in_mm_inputs() for deferred SHM cleanup.
  - request_receiver.py: skip SHM barrier+unwrap when skip_shm_flush
    is set (bg thread handles it).
  - prefill.py: rewrite event_loop_overlap_disagg_prefill into the
    pipelined structure.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #28142400155](https://github.com/sgl-project/sglang/actions/runs/28142400155)
Latest PR Test (Extra): :x: [Run #28142399918](https://github.com/sgl-project/sglang/actions/runs/28142399918)

### CODE REVIEW

File: python/sglang/srt/disaggregation/prefill.py [hunk 1]
```diff
@@ -498,13 +498,49 @@ def event_loop_normal_disagg_prefill(self: Scheduler) -> None:
 
     @torch.no_grad()
     def event_loop_overlap_disagg_prefill(self: Scheduler) -> None:
+        """Pipelined disagg-prefill event loop.
+
+        Overlaps heavy mm_inputs processing (from_processor_output + SHM
+        clone) in a background thread with GPU forward on the main thread.
+        The bg thread has NO Gloo calls, only CPU work whose GIL is released
+        during clone().
+        """
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/prefill.py]: This should not be an mm_inputs processing-related docstring, it is for disagg prefill. I prefer we don't add this here, just keep some short comments in front of the related and newly added code.

File: python/sglang/srt/disaggregation/prefill.py [hunk 2]
```diff
@@ -498,13 +498,49 @@ def event_loop_normal_disagg_prefill(self: Scheduler) -> None:
...
+            # ── 2. Recv new requests (Gloo broadcast syncs all ranks) ──
             recv_reqs = self.request_receiver.recv_requests()
-            self.process_input_requests(recv_reqs)
+            self._flush_oldest_shm_generation()
+
+            # ── 3. Submit mm_inputs processing to bg thread ──
+            prev_mm_future, prev_mm_indices = self._submit_mm_processing(
+                mm_executor, recv_reqs
+            )
+            prev_recv_reqs = recv_reqs
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/prefill.py]: This seems to affect the backbond of the loop, even without mm inputs.

Should we add a switch to let this code path only take effect on VLMs with mm_inputs?

File: python/sglang/srt/managers/scheduler_components/request_receiver.py [hunk 3]
```diff
@@ -41,7 +41,7 @@
     )
 
 
-@dataclass(kw_only=True, slots=True, frozen=True)
```
Comment (ShangmingCai) [inline python/sglang/srt/managers/scheduler_components/request_receiver.py:45]: I am not sure about removing `frozen=True`, I think mutable data causes many bugs, which could be why it is added here in the first place. Is it possible to do this in other place?

**Review by ShangmingCai — COMMENTED** [review]
Could you provide some data on the performance improvements? I feel this PR is somewhat intrusive and introduces risks related to scheduling; we need to evaluate whether the gains justify the added complexity.

**Comment by Jumiar** [comment]
> Could you provide some data on the performance improvements? I feel this PR is somewhat intrusive and introduces risks related to scheduling; we need to evaluate whether the gains justify the added complexity.

Okay, we'll add the latest results.
